xarray/core/resample.py

```diff
@@ -0,0 +1,344 @@
+                             **kwargs)
+        return self.apply(reduce_array, shortcut=shortcut)
+
+    def _interpolate(self, kind='linear'):
+        from .dataarray import DataArray
+
+        x = self._obj[self._dim].astype('float')
+        y = self._obj.values
+        axis = self._obj.get_axis_num(self._dim)
+
+        f = interp1d(x, y, kind=kind, axis=axis, bounds_error=True,
+                     assume_sorted=True)
+        new_x = self._full_index.values.astype('float')
+
+        # construct new up-sampled DataArray
+        dummy = self._obj.copy()
+        dims = dummy.dims
+        coords = dummy.coords
+        coords[self._dim] = self._full_index
+        return DataArray(f(new_x), coords, dims, name=dummy.name,
+                         attrs=dummy.attrs)
+
+ops.inject_reduce_methods(DataArrayResample)
+ops.inject_binary_ops(DataArrayResample)
+
+
+class DatasetResample(DatasetGroupBy, Resample):
+    """DatasetGroupBy object specialized to resampling a specified dimension
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        self._dim = kwargs.pop('dim', None)
+        self._resample_dim = kwargs.pop('resample_dim', None)
+
+        if self._dim == self._resample_dim:
+            raise ValueError("Proxy resampling dimension ('{_resample_dim}') "
+                             "cannot have the same name as actual dimension "
+                             "('{_dim}')! ".format(self))
+        super(DatasetResample, self).__init__(*args, **kwargs)
+
+    def apply(self, func, **kwargs):
+        """Apply a function over each Dataset in the groups generated for
+        resampling  and concatenate them together into a new Dataset.
+
+        `func` is called like `func(ds, *args, **kwargs)` for each dataset `ds`
+        in this group.
+
+        Apply uses heuristics (like `pandas.GroupBy.apply`) to figure out how
+        to stack together the datasets. The rule is:
+        1. If the dimension along which the group coordinate is defined is
+           still in the first grouped item after applying `func`, then stack
+           over this dimension.
+        2. Otherwise, stack over the new dimension given by name of this
+           grouping (the argument to the `groupby` function).
+
+        Parameters
+        ----------
+        func : function
+            Callable to apply to each sub-dataset.
+        **kwargs
+            Used to call `func(ds, **kwargs)` for each sub-dataset `ar`.
+
+        Returns
+        -------
+        applied : Dataset or DataArray
+            The result of splitting, applying and combining this dataset.
+        """
+        kwargs.pop('shortcut', None)  # ignore shortcut if set (for now)
+        applied = (func(ds, **kwargs) for ds in self._iter_grouped())
+        combined = self._combine(applied)
+
+        return combined.rename({self._resample_dim: self._dim})
+
+    def reduce(self, func, dim=None, keep_attrs=False, **kwargs):
```

Comment:
We do have to inject the **dim** argument to be `self._dim`, but I just call the super-method with that as an argument now.